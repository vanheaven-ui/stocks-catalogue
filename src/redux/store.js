import { compose, createStore } from 'redux';
import rootReducer from './reducers';

const storeEnhancers = compose(
  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__(), // eslint-disable-line
);

const store = createStore(rootReducer, storeEnhancers);

export default store;
