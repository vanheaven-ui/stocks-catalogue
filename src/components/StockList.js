import { useDispatch, useSelector } from 'react-redux';
// import PropTypes from 'prop-types';
// import useFetch from '../common/hooks/useFetch';
// import { listEndPoint } from '../constants';
import { getStockList } from '../redux/actions';
import getStocksByPrice from '../redux/selectors';
// import ApiHandler from '../utils/apiHandler';
import Stock from './Stock';
import StocksFilter from './StockFilter';

const StockList = () => {
  const dispatch = useDispatch();
  // const { data } = useFetch(listEndPoint);
  const stocks = JSON.parse(localStorage.getItem('stocks'));
  dispatch(getStockList(stocks));
  // localStorage.setItem('stocks', JSON.stringify(data));
  const filter = useSelector(state => state.filter.filter);

  // get profiles;

  const renderFilteredStocks = (() => {
    const filtered = filter === 'All' ? stocks : getStocksByPrice(stocks, filter);
    return filtered;
  })();

  return (
    <>
      <StocksFilter />
      <h2>Stock List</h2>
      { stocks && renderFilteredStocks.slice(0, 99).map(stock => (
        <article key={`stck-${stock.symbol}`}>
          <Stock stock={stock} />
        </article>
      ))}
      <p>Would be better if we can have pagination</p>
    </>
  );
};

export default StockList;
