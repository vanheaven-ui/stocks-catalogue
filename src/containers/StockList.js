import { useDispatch, useSelector } from 'react-redux';
import Container from 'react-bootstrap/Container';
import { Col, Row } from 'react-bootstrap';
import { useState } from 'react';
import ReactPaginate from 'react-paginate';
import { changeFilter } from '../redux/actions'; // , getStockList
import StocksFilter from '../components/StockFilter';
// import useFetch1 from '../common/hooks/useFetch1';
import STOCKS_FILTERS from '../constants'; // ,  { listEndPoint }
import { filterByExchange, getStocksByPrice } from '../redux/selectors';
import Stock from '../components/Stock';
import styles from '../styles/paginate.module.css';
import Loading from '../components/Loading';
import removeDups from '../utils/removeDups';
import style from '../styles/stockList.module.css';

const StockList = () => {
  const [pageNumber, setPageNumber] = useState(0);
  const stocksPerPage = 20;
  const stocksDisplayed = pageNumber * stocksPerPage;

  const dispatch = useDispatch();

  const [filter, setFilter] = useState({
    price: null,
    exchange: null,
  });

  const isLoading = false;
  // const { data, isLoading } = useFetch1(listEndPoint);

  // if (data.length > 0) {
  //   localStorage.setItem('stocks', JSON.stringify(data));
  //   dispatch(getStockList(data));
  // }

  // const stocks = useSelector((state) => state.stocks.stocks);
  const stocks = JSON.parse(localStorage.getItem('stocks'));

  const exchanges = stocks ? stocks.slice(0, 1000).map((stock) => stock.exchange) : null;
  STOCKS_FILTERS.EXCHANGE = exchanges;

  STOCKS_FILTERS.EXCHANGE = removeDups(STOCKS_FILTERS.EXCHANGE);

  const handleChange = (e) => {
    if (e.target.name === 'PRICE') {
      setFilter({ ...filter, price: e.target.value });
    } if (e.target.name === 'EXCHANGE') {
      setFilter({ ...filter, exchange: e.target.value });
    }
    return 'default';
  };

  const chosenValue = (() => {
    if (filter.price) {
      return filter.price;
    } if (filter.exchange) {
      return filter.exchange;
    }
    return 'All';
  })();

  const handleSubmit = (e) => {
    e.preventDefault();
    dispatch(changeFilter(filter));
    setFilter('default');
  };

  const updateFilter = useSelector((state) => state.filter.filter);

  const renderFilteredStocks = (() => {
    let stocks2render = stocks.slice(0, 1000);
    if (updateFilter.price) {
      stocks2render = getStocksByPrice(stocks2render, updateFilter.price);
    } if (updateFilter.exchange) {
      stocks2render = filterByExchange(stocks2render, updateFilter.exchange);
    }
    return stocks2render;
  })();

  const sliceStart = stocksDisplayed;
  const sliceEnd = stocksDisplayed + stocksPerPage;
  const displayStocksPerPage = renderFilteredStocks.slice(sliceStart, sliceEnd);

  const pageCount = stocks.slice(0, 1000).length / stocksPerPage;

  const pageChange = ({ selected }) => {
    setPageNumber(selected);
  };

  return (
    <section className="stock-list">
      <Container fluid className={style.listContainer}>
        <StocksFilter
          handleChange={handleChange}
          handleSubmit={handleSubmit}
          chosenValue={chosenValue}
        />
        { isLoading ? <Loading color="#d1450d" /> : (
          <>
            <Row className="px-2">
              { displayStocksPerPage.length < 1
                && (
                <Col style={{
                  textAlign: 'center', color: '#ff3300', fontSize: '1.3rem', padding: '16px 0',
                }}
                >
                  No stocks in this criterion, hit apply to get all stocks
                  {' '}
                  <span style={{ color: 'blue' }}>or</span>
                  {' '}
                  change criterion
                </Col>
                )}
              { displayStocksPerPage.length > 0 && displayStocksPerPage.map((stock) => (
                <Col sm={12} md={6} lg={3} key={`stck-${stock.symbol}`} className="px-2">
                  <article
                    className={style.stockPreview}
                  >
                    <Stock stock={stock} />
                  </article>
                </Col>
              ))}
            </Row>
            <ReactPaginate
              previousLabel="PREV"
              nextLabel="NEXT"
              pageCount={pageCount}
              onPageChange={pageChange}
              containerClassName={styles.myClass}
              previousLinkClassName={styles.prev}
              nextLinkClassName={styles.next}
              activeClassName={styles.myActive}
            />
          </>
        )}

      </Container>
    </section>
  );
};

export default StockList;
