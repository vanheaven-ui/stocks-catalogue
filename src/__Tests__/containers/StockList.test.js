import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { cleanup, render } from '@testing-library/react';
import StockList from '../../containers/StockList';
import store from '../../redux/store';
import rootNode from '../helpers';
import '@testing-library/jest-dom';

describe('StockList container component', () => {
  afterEach(cleanup);

  test('renders without crashing', () => {
    ReactDOM.render(<Provider store={store}><StockList /></Provider>, rootNode());
  });

  test('doesnot change unexpectedly', () => {
    const tree = renderer.create(
      <Provider store={store}><StockList /></Provider>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('renders loading component when getting data', () => {
    const { getByText } = render(<Provider store={store}><StockList /></Provider>);
    expect(getByText(/Loading.../)).toBeInTheDocument();
  });
});
