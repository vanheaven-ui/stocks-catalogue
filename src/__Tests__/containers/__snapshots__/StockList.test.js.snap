// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StockList container component doesnot change unexpectedly 1`] = `
<section
  className="stock-list"
>
  <div
    className="listContainer container-fluid"
  >
    <section
      className="filters"
    >
      <h2>
        Browse Stocks with Filters:
      </h2>
      <div
        className="container"
      >
        <form
          onSubmit={[Function]}
          style={
            Object {
              "textAlign": "center",
            }
          }
        >
          <div
            className="filters"
          >
            <select
              name="PRICE"
              onChange={[Function]}
              value="All"
            >
              <option
                value="default"
              >
                Select:
                 
                PRICE RANGE
              </option>
              <option
                value="<100"
              >
                &lt;100
              </option>
              <option
                value="100 < Price < 200"
              >
                100 &lt; Price &lt; 200
              </option>
              <option
                value="200 < Price < 300"
              >
                200 &lt; Price &lt; 300
              </option>
              <option
                value="300 < Price < 400"
              >
                300 &lt; Price &lt; 400
              </option>
              <option
                value="400 < Price < 500"
              >
                400 &lt; Price &lt; 500
              </option>
              <option
                value=">500"
              >
                &gt;500
              </option>
            </select>
            <select
              name="EXCHANGE"
              onChange={[Function]}
              value="All"
            >
              <option
                value="default"
              >
                Select:
                 
                EXCHANGE
              </option>
            </select>
            <button
              type="submit"
            >
              Apply
            </button>
          </div>
        </form>
      </div>
    </section>
    <div
      className="px-2 row"
    >
      <div
        className="col"
        style={
          Object {
            "color": "#ff3300",
            "fontSize": "1.3rem",
            "padding": "16px 0",
            "textAlign": "center",
          }
        }
      >
        No stocks in this criterion, hit apply to get all stocks
         
        <span
          style={
            Object {
              "color": "blue",
            }
          }
        >
          or
        </span>
         
        change criterion
      </div>
    </div>
    <ul
      className="myClass"
    >
      <li
        className="previous disabled"
      >
        <a
          aria-disabled="true"
          aria-label="Previous page"
          className="prev"
          onClick={[Function]}
          onKeyPress={[Function]}
          rel="prev"
          role="button"
          tabIndex="0"
        >
          PREV
        </a>
      </li>
      <li
        className="next"
      >
        <a
          aria-disabled="false"
          aria-label="Next page"
          className="next"
          onClick={[Function]}
          onKeyPress={[Function]}
          rel="next"
          role="button"
          tabIndex="0"
        >
          NEXT
        </a>
      </li>
    </ul>
  </div>
</section>
`;
